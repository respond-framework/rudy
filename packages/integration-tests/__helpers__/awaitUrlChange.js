// `await pop() and await snapPop` do something similar to below (see `createTest.js`).
// However we need a manual one that checks if the URL has changed as well, to insure
// that actions dispatched after a pop come after the browser would have handled the
// pop, which is asynchronous in some browsers (chrome), as well as Jest. In real life
// the user takes at lesat 65ms to pop again. In code, it's too fast at 0ms. So we
// wait a few more ms for it to actually change and trigger our pop handling code.

/* eslint-env browser */

export default () => {
  const currentUrl = window.location.href
  return haschanged(currentUrl)
}

const haschanged = async (currentUrl, tries = 1) => {
  if (tries >= 10) {
    throw new Error('awaitUrlChange reached the maximum amount of tries (10)')
  }

  await new Promise((res) => setTimeout(res, 5))
  if (currentUrl !== window.location.href) return
  return haschanged(currentUrl, ++tries)
}

export const windowHistoryGo = (n) =>
  new Promise((resolve, reject) => {
    let timeout

    // Resolve when the resulting popstate event happens
    const onPopState = () => {
      clearTimeout(timeout)
      window.removeEventListener('popstate', onPopState)
      resolve()
    }
    window.addEventListener('popstate', onPopState)

    // Trigger the real go function
    window.history.go(n)

    // timeout after 1 second
    timeout = setTimeout(() => {
      window.removeEventListener('popstate', onPopState)
      reject()
    }, 1000)
  })
