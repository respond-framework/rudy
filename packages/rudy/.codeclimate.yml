languages:
  JavaScript: true

engines:
  duplication:
    enabled: true
    config:
      languages:
        - javascript:
  fixme:
    enabled: true
  eslint:
    enabled: true
    config:
      config: .eslintrc.js
    checks:
      import/no-unresolved:
        enabled: false
      import/extensions:
        enabled: false

ratings:
  paths:
    - 'src/**'

exclude_paths:
  - 'docs/'
  - 'dist/'
  - 'flow-typed/'
  - 'node_modules/'
  - '.vscode/'
  - '.eslintrc.js'
  - '**/*.snap'
