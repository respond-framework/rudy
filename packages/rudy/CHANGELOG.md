# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.1-from-git.1](https://github.com/respond-framework/rudy/tree/master/packages/rudy/compare/@respond-framework/rudy@0.1.1-test.3...@respond-framework/rudy@0.1.1-from-git.1) (2019-08-21)

**Note:** Version bump only for package @respond-framework/rudy





## [0.1.1-from-git.0](https://github.com/respond-framework/rudy/tree/master/packages/rudy/compare/@respond-framework/rudy@0.1.1-test.3...@respond-framework/rudy@0.1.1-from-git.0) (2019-08-21)

**Note:** Version bump only for package @respond-framework/rudy





## [0.1.1-test.3](https://github.com/respond-framework/rudy/tree/master/packages/rudy/compare/@respond-framework/rudy@0.1.1-test.2...@respond-framework/rudy@0.1.1-test.3) (2019-06-07)


### Build System

* add ability to publish to git instead of NPM ([#34](https://github.com/respond-framework/rudy/tree/master/packages/rudy/issues/34)) ([95e59e0](https://github.com/respond-framework/rudy/tree/master/packages/rudy/commit/95e59e0))


### Features

* add compatibility with react-redux 6+ ([#33](https://github.com/respond-framework/rudy/tree/master/packages/rudy/issues/33)) ([cb45152](https://github.com/respond-framework/rudy/tree/master/packages/rudy/commit/cb45152))


### BREAKING CHANGES

* The object returned from createRouter now contains a key `api` which contains `routes`, `ctx` and such things. Previously they were in the root object.
* The `link` package has been renamed to `react`.
* The `Link` component now must be rendered as a descendant of the new `RudyProvider`.
* The core no longer monkey patches redux `getState` to have a key `api` with the rudy API.
* Remove UMD builds





## [0.1.1-test.2](https://github.com/respond-framework/rudy/tree/master/packages/rudy/compare/@respond-framework/rudy@0.1.1-test.1...@respond-framework/rudy@0.1.1-test.2) (2019-06-06)


### Bug Fixes

* **core:** Pass the configured title state key or selector into the title middleware [#28](https://github.com/respond-framework/rudy/tree/master/packages/rudy/issues/28) ([381d8c3](https://github.com/respond-framework/rudy/tree/master/packages/rudy/commit/381d8c3))





## 0.1.1-test.1 (2019-04-15)


### Features

* add options to the call middleware ([#17](https://github.com/respond-framework/rudy/tree/master/packages/rudy/issues/17)) ([4ee0198](https://github.com/respond-framework/rudy/tree/master/packages/rudy/commit/4ee0198))
* **transformations:** Various changes, documentation ([#24](https://github.com/respond-framework/rudy/tree/master/packages/rudy/issues/24)) ([c7c2320](https://github.com/respond-framework/rudy/tree/master/packages/rudy/commit/c7c2320))


### BREAKING CHANGES

* **transformations:** `fromPath` and `toPath` Have changed type signatures, and the defaults behave differently.





## 0.1.1-test.0 (2018-11-05)

**Note:** Version bump only for package @respond-framework/rudy
