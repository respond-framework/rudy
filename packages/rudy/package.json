{
  "name": "@respond-framework/rudy",
  "version": "0.1.1-from-git.1",
  "description": "think of your app in states not routes (and, yes, while keeping the address bar in sync)",
  "main": "dist/index.js",
  "module": "es/index.js",
  "rudy-src-main": "src/index.js",
  "files": [
    "dist",
    "es"
  ],
  "scripts": {
    "prepare": "yarn run build:cjs && yarn run build:es && yarn run flow-copy",
    "build:cjs": "babel --root-mode upward --source-maps true src -d dist",
    "build:es": "BABEL_ENV=es babel --root-mode upward --source-maps true src -d es",
    "build": "yarn run build:cjs && yarn run build:es",
    "flow-copy": "flow-copy-source src dist && flow-copy-source src es",
    "flow-watch": "clear; printf \"\\033[3J\" & npm run flow & fswatch -o ./ | xargs -n1 -I{} sh -c 'clear; printf \"\\033[3J\" && npm run flow'",
    "flow": "flow; test $? -eq 0 -o $? -eq 2",
    "clean": "rimraf dist && rimraf es && rimraf coverage",
    "prettier": "prettier",
    "is-pretty": "prettier --ignore-path=../../config/.prettierignore '**/*' --list-different",
    "prettify": "prettier --ignore-path=../../config/.prettierignore '**/*' --write",
    "lint": "eslint . || true",
    "test": "jest --config ../../jest.config.js --rootDir ."
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/respond-framework/rudy/tree/master/packages/rudy"
  },
  "author": "James Gillmore <james@faceyspacey.com>",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "bugs": {
    "url": "https://github.com/respond-framework/rudy/issues"
  },
  "homepage": "https://github.com/respond-framework/rudy",
  "dependencies": {
    "@respond-framework/middleware-change-page-title": "^1.0.1-test.1",
    "@respond-framework/utils": "^0.1.1-test.1",
    "path-to-regexp": "^2.1.0",
    "prop-types": "^15.6.0",
    "qs": "^6.5.1",
    "resolve-pathname": "^2.2.0"
  },
  "lint-staged": {
    "*.js": [
      "prettier --single-quote --parser=flow --semi=false --write",
      "eslint --fix",
      "git add"
    ]
  }
}
