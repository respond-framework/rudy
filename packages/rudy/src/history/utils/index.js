export { supportsDom } from './supports'

export {
  addPopListener,
  removePopListener,
  isExtraneousPopEvent,
} from './popListener'

export {
  saveHistory,
  restoreHistory,
  pushState,
  replaceState,
  go,
  getCurrentIndex,
} from './sessionStorage'
