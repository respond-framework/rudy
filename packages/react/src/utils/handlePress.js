// @flow
import { redirect } from '@respond-framework/rudy'
import type { Routes, ReceivedAction } from '@respond-framework/rudy'

export type OnClick = false | ((SyntheticEvent) => ?boolean)
export default (
  action: ?ReceivedAction,
  routes: Routes,
  shouldDispatch: boolean,
  dispatch: Function,
  onClick?: ?OnClick,
  target: ?string,
  isRedirect?: boolean,
  fullUrl: string,
  history: Object,
  e: SyntheticEvent,
) => {
  const prevented = e.defaultPrevented
  const notModified = !isModified(e)
  let shouldGo = true

  if (onClick) {
    shouldGo = onClick(e) !== false // onClick can return false to prevent dispatch
  }

  if (!target && e && e.preventDefault && notModified) {
    e.preventDefault()
  }

  if (
    action &&
    shouldGo &&
    shouldDispatch &&
    !target &&
    !prevented &&
    notModified &&
    e.button === 0
  ) {
    action = isRedirect ? redirect(action) : action
    return dispatch(action)
  }

  if (!action && !target && fullUrl.indexOf('http') === 0) {
    window.location.href = fullUrl
  }
  return undefined
}

const isModified = (e: Object) =>
  !!(e.metaKey || e.altKey || e.ctrlKey || e.shiftKey)
