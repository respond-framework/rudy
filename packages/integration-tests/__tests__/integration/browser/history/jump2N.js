import { locationToUrl } from '@respond-framework/rudy/utils'
import { jump } from '@respond-framework/rudy/actions'
import createTest, { resetBrowser } from '../../../../__helpers__/createTest'
import { windowHistoryGo } from '../../../../__helpers__/awaitUrlChange'

beforeEach(resetBrowser)

createTest(
  'set(action, n)',
  {
    SECOND: '/second',
    THIRD: '/third',
    FIRST: '/:foo?',
  },
  {
    testBrowser: true,
  },
  [],
  async ({ dispatch, snap, snapPop }) => {
    expect(locationToUrl(window.location)).toEqual('/')

    await dispatch({ type: 'SECOND' })
    await dispatch({ type: 'THIRD' })

    expect(locationToUrl(window.location)).toEqual('/third')

    await snap(jump(-2))

    expect(locationToUrl(window.location)).toEqual('/')

    await snapPop('forward')
    expect(locationToUrl(window.location)).toEqual('/second')

    await snapPop('back')

    await dispatch(jump(2))
    expect(locationToUrl(window.location)).toEqual('/third')
  },
)

createTest(
  'history.go(n)',
  {
    SECOND: '/second',
    THIRD: '/third',
    FIRST: '/:foo?',
  },
  {
    testBrowser: true,
  },
  [],
  async ({ dispatch, snap }) => {
    expect(locationToUrl(window.location)).toEqual('/')

    await dispatch({ type: 'SECOND' })
    await dispatch({ type: 'THIRD' })

    expect(locationToUrl(window.location)).toEqual('/third')

    // This simulates what happens if the user right clicks on the back
    // button, and goes back by two steps
    await windowHistoryGo(-2)
    expect(locationToUrl(window.location)).toEqual('/')
    await snap()

    await windowHistoryGo(1)
    expect(locationToUrl(window.location)).toEqual('/second')
    await snap()

    await windowHistoryGo(-1)
    expect(locationToUrl(window.location)).toEqual('/')
    await snap()

    await windowHistoryGo(2)
    expect(locationToUrl(window.location)).toEqual('/third')
    await snap()
  },
)
