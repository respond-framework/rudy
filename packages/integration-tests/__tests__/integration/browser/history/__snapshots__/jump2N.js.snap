// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`history.go(n): <none> - response 1`] = `null`;

exports[`history.go(n): <none> - response 2`] = `null`;

exports[`history.go(n): <none> - response 3`] = `null`;

exports[`history.go(n): <none> - response 4`] = `null`;

exports[`history.go(n): <none> - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`history.go(n): <none> - routes - NEVER_USED_PATHLESS - thunk 2`] = `0`;

exports[`history.go(n): <none> - routes - NEVER_USED_PATHLESS - thunk 3`] = `0`;

exports[`history.go(n): <none> - routes - NEVER_USED_PATHLESS - thunk 4`] = `0`;

exports[`history.go(n): <none> - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`history.go(n): <none> - routes - REDIRECTED - onComplete 2`] = `0`;

exports[`history.go(n): <none> - routes - REDIRECTED - onComplete 3`] = `0`;

exports[`history.go(n): <none> - routes - REDIRECTED - onComplete 4`] = `0`;

exports[`history.go(n): <none> - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`history.go(n): <none> - sessionStorage 2`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`history.go(n): <none> - sessionStorage 3`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`history.go(n): <none> - sessionStorage 4`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`history.go(n): <none> - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "jump",
    "length": 3,
    "n": -2,
    "params": Object {},
    "pathname": "/",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 2,
        "key": "345678",
        "pathname": "/third",
        "scene": "",
        "search": "",
        "url": "/third",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "THIRD",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`history.go(n): <none> - state 2`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 1,
    "key": "345678",
    "kind": "next",
    "length": 3,
    "n": 1,
    "params": Object {},
    "pathname": "/second",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "SECOND",
    "universal": false,
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`history.go(n): <none> - state 3`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "back",
    "length": 3,
    "n": -1,
    "params": Object {},
    "pathname": "/",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`history.go(n): <none> - state 4`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 2,
    "key": "345678",
    "kind": "jump",
    "length": 3,
    "n": 2,
    "params": Object {},
    "pathname": "/third",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "THIRD",
    "universal": false,
    "url": "/third",
  },
  "title": "THIRD",
}
`;

exports[`history.go(n): <none> - title 1`] = `"THIRD"`;

exports[`history.go(n): <none> - title 2`] = `"FIRST"`;

exports[`history.go(n): <none> - title 3`] = `"SECOND"`;

exports[`history.go(n): <none> - title 4`] = `"FIRST"`;

exports[`history.go(n): <none> - windowLocation 1`] = `"/"`;

exports[`history.go(n): <none> - windowLocation 2`] = `"/second"`;

exports[`history.go(n): <none> - windowLocation 3`] = `"/"`;

exports[`history.go(n): <none> - windowLocation 4`] = `"/third"`;

exports[`set(action, n):  - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "index": 1,
    "key": "345678",
    "kind": "next",
    "length": 3,
    "n": 1,
    "pathname": "/second",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "SECOND",
}
`;

exports[`set(action, n):  - response 2`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "index": 0,
    "key": "345678",
    "kind": "back",
    "length": 3,
    "n": -1,
    "pathname": "/",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`set(action, n):  - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`set(action, n):  - routes - NEVER_USED_PATHLESS - thunk 2`] = `0`;

exports[`set(action, n):  - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`set(action, n):  - routes - REDIRECTED - onComplete 2`] = `0`;

exports[`set(action, n):  - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`set(action, n):  - sessionStorage 2`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`set(action, n):  - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 1,
    "key": "345678",
    "kind": "next",
    "length": 3,
    "n": 1,
    "params": Object {},
    "pathname": "/second",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "SECOND",
    "universal": false,
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`set(action, n):  - state 2`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "back",
    "length": 3,
    "n": -1,
    "params": Object {},
    "pathname": "/",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`set(action, n):  - title 1`] = `"SECOND"`;

exports[`set(action, n):  - title 2`] = `"FIRST"`;

exports[`set(action, n):  - windowLocation 1`] = `"/second"`;

exports[`set(action, n):  - windowLocation 2`] = `"/"`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "index": 0,
    "key": "345678",
    "kind": "jump",
    "length": 3,
    "n": -1,
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
    Array [
      "/third",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/third",
          "scene": "",
          "search": "",
          "url": "/third",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "THIRD",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "jump",
    "length": 3,
    "n": -1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - title 1`] = `"FIRST"`;

exports[`set(action, n): {"type":"@@rudy/CALL_HISTORY","payload":{"method":"jump","args":[-2,null,null]}} - windowLocation 1`] = `"/"`;
